pkg/csi/service/vanilla/controller.go

```diff
@@ -1137,15 +1139,155 @@ func (c *controller) CreateSnapshot(ctx context.Context, req *csi.CreateSnapshot
 	ctx = logger.NewContextWithLogger(ctx)
 	log := logger.GetLogger(ctx)
 	log.Infof("CreateSnapshot: called with args %+v", *req)
-	return nil, logger.LogNewErrorCode(log, codes.Unimplemented, "createSnapshot")
+
+	isBlockVolumeSnapshotEnabled := commonco.ContainerOrchestratorUtility.IsFSSEnabled(ctx, common.BlockVolumeSnapshot)
+	if !isBlockVolumeSnapshotEnabled {
+		return nil, logger.LogNewErrorCode(log, codes.Unimplemented, "createSnapshot")
+	}
+
+	volumeType := prometheus.PrometheusUnknownVolumeType
+	createSnapshotInternal := func() (*csi.CreateSnapshotResponse, error) {
+		// Validate CreateSnapshotRequest
+		if err := validateVanillaCreateSnapshotRequestRequest(ctx, req); err != nil {
+			return nil, logger.LogNewErrorCodef(log, codes.Internal,
+				"validation for CreateSnapshot Request: %+v has failed. Error: %v", *req, err)
+		}
+		volumeID := req.GetSourceVolumeId()
+
+		// Check if the source volume is migrated vSphere volume
+		if strings.Contains(volumeID, ".vmdk") {
+			return nil, logger.LogNewErrorCodef(log, codes.Unimplemented,
+				"cannot snapshot migrated vSphere volume. :%q", volumeID)
+		}
+
+		// Check if volume already exists
+		volumeIds := []cnstypes.CnsVolumeId{{Id: volumeID}}
+		queryFilter := cnstypes.CnsQueryFilter{
+			VolumeIds: volumeIds,
+		}
+		queryResult, err := c.manager.VolumeManager.QueryVolume(ctx, queryFilter)
+		if err != nil {
+			return nil, logger.LogNewErrorCodef(log, codes.Internal,
+				"The volume %q to be snapshotted is not available: %v", volumeID, err)
+		}
+
+		// Check if it is a block volume
+		if queryResult.Volumes[0].VolumeType != common.BlockVolumeType {
+			return nil, logger.LogNewErrorCodef(log, codes.Unimplemented,
+				"CreateSnapshot is only supported on block volume. VolumeType: %v",
+				queryResult.Volumes[0].VolumeType)
+		}
+		volumeType = prometheus.PrometheusBlockVolumeType
+
+		// Get the snapshot size by getting the volume size, which is only valid for full backup use cases
+		var snapshotSize int64
+		if len(queryResult.Volumes) > 0 {
+			snapshotSize = queryResult.Volumes[0].BackingObjectDetails.GetCnsBackingObjectDetails().CapacityInMb
+		} else {
+			return nil, logger.LogNewErrorCodef(log, codes.Internal,
+				"failed to get the snapshot size by querying volume: %q", volumeID)
+		}
+
+		// the returned snapshotID below is a combination of CNS VolumeID and CNS SnapshotID concatenated by the "+"
+		// sign. That is, a string of "<UUID>+<UUID>"
```

Comment:
Add a comment that this is because we need volume id when calling delete snapshot CNS API.